src/server/electronLoader.ts

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as electron from 'electron';
+
+// Electron loader is passed to Electron as --require electronLoader.js and
+// defers the app ready event until remote debugging is established.
+
+(async () => {
+  const app = electron.app;
+  const originalEmitMethod = app.emit.bind(app);
+  const originalIsReadyMethod = app.isReady.bind(app);
+  const originalWhenReadyMethod = app.whenReady.bind(app);
+
+  const deferredEmits: { event: string | symbol, args: any[] }[] = [];
+  let attached = false;
+  let isReady = false;
+  let readyCallback: () => void;
+  const readyPromise = new Promise(f => readyCallback = f);
+
+  app.isReady = () => {
+    if (attached)
+      return originalIsReadyMethod();
+    return isReady;
+  };
+
+  app.whenReady = async () => {
```

Comment:
Same here: `await readyPromise; await originalWhenReadyMethod();`

This way, you don't have to special-case ready event. Overall, I'd make it something along these lines:
```
let attached = false;
let attachedCallback;
const attachedPromise = new Promise(f => attachedCallback = f);

app.isReady = () => attached && originalIsReadyMethod();
app.whenReady = async () => {
  await attachedPromise;
  await originalWhenReadyMethod();
};
app.emit = (event, ...args) => {
  if (attached)
    return originalEmitMethod(event, ...args);
  attachedPromise.then(() => originalEmitMethod(event, ...args));
  return true;
};

const playwrightRun = () => {
  attachedCallback();
  return Promise.resolve(electron);
};
```

Reply:
Discussed offline how isReady needs to correlate w/ message flow.